Systematic `panic: runtime error: index out of range [-1]` while building container

<!---
Thank you for contributing to BuildKit through issue submission! If you're new to this repository, we encourage you to review our issue reporting guide https://github.com/moby/buildkit/blob/master/.github/issue_reporting_guide.md which outlines the key information you should provide for the process to go smoothly.

Feel free to report bugs, suggest features, or submit proposals here. For general questions about using BuildKit, consider joining the #buildkit channel on the Docker Community Slack, where a broader community can provide support and insights.
--->

Hi team, I'm seeing this stack trace in `~/Library/Containers/com.docker.docker/Data/log/vm/dockerd.log` while trying to build a dockerfile. I'm not sure what's causing it, and unfortunately I'm not able to share the dockerfile.

Platform:
* Apple M1 Pro, Sonoma 14.5
* Docker desktop version 4.31.0 (153195) - but I also saw it on 4.32.0

```
panic: runtime error: index out of range [-1]

goroutine 112437 [running]:
github.com/moby/buildkit/frontend/dockerfile/dockerfile2llb.(*dispatchStates).lastTarget(...)
	/root/build-deb/engine/vendor/github.com/moby/buildkit/frontend/dockerfile/dockerfile2llb/convert.go:904
github.com/moby/buildkit/frontend/dockerfile/dockerfile2llb.toDispatchState({_, _}, {_, _, _}, {{0x4002a23b90, {0x0, 0x0}, {0x0, 0x0}, ...}, ...})
	/root/build-deb/engine/vendor/github.com/moby/buildkit/frontend/dockerfile/dockerfile2llb/convert.go:323 +0x3278
github.com/moby/buildkit/frontend/dockerfile/dockerfile2llb.Dockerfile2LLB({_, _}, {_, _, _}, {{0x4002a23b90, {0x0, 0x0}, {0x0, 0x0}, ...}, ...})
	/root/build-deb/engine/vendor/github.com/moby/buildkit/frontend/dockerfile/dockerfile2llb/convert.go:76 +0x54
github.com/moby/buildkit/frontend/dockerfile/builder.Build.func5({0xaaaae234e118, 0x4003053540}, 0x4002066cc0, 0x400147dc98?)
	/root/build-deb/engine/vendor/github.com/moby/buildkit/frontend/dockerfile/builder/build.go:125 +0xc8
github.com/moby/buildkit/frontend/dockerui.(*Client).Build.func1()
	/root/build-deb/engine/vendor/github.com/moby/buildkit/frontend/dockerui/build.go:39 +0x64
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/root/build-deb/engine/vendor/golang.org/x/sync/errgroup/errgroup.go:75 +0x58
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 112386
	/root/build-deb/engine/vendor/golang.org/x/sync/errgroup/errgroup.go:72 +0x98
EOF
```

--- a/frontend/dockerfile/dockerfile2llb/convert.go
+++ b/frontend/dockerfile/dockerfile2llb/convert.go
@@ -258,6 +258,9 @@ func toDispatchState(ctx context.Context, dt []byte, opt ConvertOpt) (*dispatchS
 	if err != nil {
 		return nil, err
 	}
+	if len(stages) == 0 {
+		return nil, errors.New("dockerfile contains no stages to build")
+	}
 	validateStageNames(stages, lint)
 
 	shlex := shell.NewLex(dockerfile.EscapeToken)

--- a/frontend/dockerfile/dockerfile_test.go
+++ b/frontend/dockerfile/dockerfile_test.go
@@ -196,6 +196,7 @@ var allTests = integration.TestFuncs(
 	testInvalidJSONCommands,
 	testHistoryError,
 	testHistoryFinalizeTrace,
+	testEmptyStages,
 )
 
 // Tests that depend on the `security.*` entitlements
@@ -5477,6 +5478,40 @@ RUN echo $(hostname) | grep foo
 			require.NoError(t, err)
 		})
 	}
+}
+
+func testEmptyStages(t *testing.T, sb integration.Sandbox) {
+	integration.SkipOnPlatform(t, "windows")
+	f := getFrontend(t, sb)
+	dockerfile := []byte(`
+ARG foo=bar
+`)
+
+	dir := integration.Tmpdir(
+		t,
+		fstest.CreateFile("Dockerfile", dockerfile, 0600),
+	)
+
+	c, err := client.New(sb.Context(), sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	destDir := t.TempDir()
+
+	_, err = f.Solve(sb.Context(), c, client.SolveOpt{
+		LocalMounts: map[string]fsutil.FS{
+			dockerui.DefaultLocalNameDockerfile: dir,
+			dockerui.DefaultLocalNameContext:    dir,
+		},
+		Exports: []client.ExportEntry{
+			{
+				Type:      client.ExporterLocal,
+				OutputDir: destDir,
+			},
+		},
+	}, nil)
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "dockerfile contains no stages to build")
 }
 
 func testShmSize(t *testing.T, sb integration.Sandbox) {

